_blog/2023-12-20-loco-rust-rails.mdx

```diff
@@ -0,0 +1,219 @@
+
+## Getting Started
+To get started with Loco, you need to use their CLI which you can install by using the following:
+```bash
+cargo install loco-cli
+```
+
+You can start a new project by using `loco new` - it'll ask you what the name of your app is and then what kind of app you want. For this article, we'll be talking about the full SaaS starter application.
+
+## Routing in Loco
+Although Loco uses axum under the hood, it abstracts some things away into config files that you can find in the `configuration` folder. 
+
+The Axum service that gets run by the application implements the `Hooks` trait from `loco_cli`, which requires several functions to use - going to `src/app.rs` shows that we have functions for registering routes, getting the app name, connecting workers and registering tasks, truncate tables and seed data into the database. We can also add extra functions to the router that get hooked into the CLI as well - `after_routes()` which is for adding things like middleware, and `before_run()` which allows you to carry out operations before your application itself starts. Note that any commands we use through the project CLI to generate things will automatically be appended to the `app.rs` file - no need to do it yourself!
+
+To add a controller, we need to run `cargo run generate controller test` from the project root which generates a controller called `test` and simultaneously adds a new file in the `controllers` folder. Then we can create any routes we need to and append them to the router in the same file, and it'll automatically be added to the application - no further work required! Your new controller should look like something like this:
+```rust
+#![allow(clippy::unused_async)]
+use loco_rs::prelude::*;
+
+pub async fn echo(req_body: String) -> String {
+    req_body
+}
+
+pub async fn hello(State(_ctx): State<AppContext>) -> Result<String> {
+    // do something with context (database, etc)
+    format::text("hello")
+}
+
+pub fn routes() -> Routes {
+    Routes::new()
+        .prefix("test")
+        .add("/", get(hello))
+        .add("/echo", post(echo))
+}
+```
+Now you can add any routes you want to this file and it will be put under this controller when it's added to the `routes()` function in the file. Route-wise, this means it'll be all under the same route. You can access the database connection from the provided `State` - unlike in Axum normally, you don't need to create this yourself.
+
+The great thing about Loco's routing is that everything you know from using Axum can be applied here - so if you know how to write your own extractors, write middleware, and other things this can all be used in Loco since it essentially builds on top of Axum. 
+
+Once you've finished adding all the controllers and routes you want, you can use `cargo run routes` to display all of the routes your application currently has.
+
+## Migrations
```

Comment:
to make things a bit more lightweight, you can skip this -- generating a model generates migrations automatically :)

migrations are used only when you need to modify tables, add index, etc.